lib/elixir/lib/module.ex

```
@@ -1097,14 +1106,120 @@ defmodule Module do
+  defp format_function(name, arity)
+       when (is_atom(name) or is_binary(name)) and is_integer(arity) do
+    "#{name}/#{arity}"
+  end
+
+  defp do_check_impl(_, kind, _, value, _)
+       when value == nil and kind in [:type, :typep, :defp, :defmacrop] do
+    :ok
+  end
+  defp do_check_impl(_, kind, function_tuple, _, _) when kind in [:defp, :defmacrop] do
+    {:error, "function #{format_function(function_tuple)} is private, @impl " <>
+             "is always ignored for private functions"}
+  end
+  defp do_check_impl(_, kind, _, _, _) when kind in [:type, :typep] do
+    {:error, "@impl is always ignored for type annotations"}
+  end
+  defp do_check_impl(_, _, _, value, impl_required_for_callbacks?)
+       when value in [nil, false] and not impl_required_for_callbacks? do
+    :ok
+  end
+  defp do_check_impl(module, _, function_tuple, value, impl_required_for_callbacks?)
+       when value in [nil, false] and impl_required_for_callbacks? do
+    callbacks_with_behaviour = for behaviour <- get_behaviours(module),
+                                   callback <- behaviour.behaviour_info(:callbacks),
+                                   into: %{},
+                                   do: {callback, behaviour}
+
+    if %{^function_tuple => behaviour} = callbacks_with_behaviour do
+      message = "module attribute @impl was not set for implemented callback " <>
+                "#{format_function(function_tuple)} (callback specified in " <>
+                "#{inspect(behaviour)})"
+      {:error, message}
+    else
+      :ok
+    end
+  end
+  defp do_check_impl(_module, _kind, _function_tuple, value, _) when value == true do
+    raise "Implement me"
+  end
+  defp do_check_impl(_module, _kind, function_tuple, behaviour, _) when is_atom(behaviour) do
+    cond do
+      not Code.ensure_compiled?(behaviour) ->
+        {:error, "got @impl #{inspect(behaviour)} but #{inspect(behaviour)} " <>
```

Perhaps we should actually raise instead of warn here?